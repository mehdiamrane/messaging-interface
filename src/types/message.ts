export interface Message {
  id: number;
  conversationId: number;
  authorId: number;
  timestamp: number;
  body: string;
}

export interface MessageSent {
  conversationId: number;
  authorId: number;
  timestamp: number;
  body: string;
}
