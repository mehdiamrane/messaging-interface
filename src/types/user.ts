export interface User {
  id: number;
  nickname: string;
  token: string;
}

export type BareUser = Omit<User, 'token'>;
