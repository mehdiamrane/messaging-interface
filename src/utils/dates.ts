import formatDistanceToNow from 'date-fns/formatDistanceToNow';
import intlFormat from 'date-fns/intlFormat';
import { enUS, fr } from 'date-fns/locale';

const localeMapping = { 'en-US': enUS, fr };

export const formatRelativeDate = (date: Date, locale: string): string =>
  formatDistanceToNow(date, { locale: localeMapping[locale], addSuffix: true });

export const formatDate = (date: Date, locale: string): string =>
  intlFormat(
    date,
    {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
    },
    {
      locale,
    },
  );

export const timestampToDate = (timestamp: number): Date => new Date(timestamp * 1000);
