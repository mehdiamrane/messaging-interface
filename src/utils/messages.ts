import type { Message } from 'src/types/message';
import type { ChatBubbleProps } from 'src/components/conversation/chatBubble/ChatBubble';

export const sortMessages = (messages: Message[]): Message[] =>
  (messages || []).sort((a, b) => a.timestamp - b.timestamp);

export const getMessagePosition = (messages: Message[], index: number): ChatBubbleProps['position'] => {
  const previous = messages[index - 1];
  const current = messages[index];
  const next = messages[index + 1];
  if (previous?.authorId !== current.authorId && next?.authorId !== current.authorId) {
    return 'single';
  }
  if (previous?.authorId !== current.authorId && next?.authorId === current.authorId) {
    return 'first';
  }
  if (previous?.authorId === current.authorId && next?.authorId !== current.authorId) {
    return 'last';
  }
  return 'middle';
};
