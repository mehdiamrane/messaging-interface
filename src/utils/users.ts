import type { User } from 'src/types/user';

// Default way to use a logged user
export const getLoggedUserId = (): User['id'] => 1;

export const isLoggedUser = (userId: User['id']): boolean => userId === getLoggedUserId();

export const getUserData = (userId: User['id'], users: User[]): User | undefined =>
  users.find(({ id }) => id === userId);

export const getAvatarFromNickname = (nickname: User['nickname']): string =>
  `https://ui-avatars.com/api/?bold=true&size=54&background=CBD5E0&rounded=true&name=${nickname}`;
