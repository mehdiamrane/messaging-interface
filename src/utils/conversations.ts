import type { Conversation } from 'src/types/conversation';
import type { BareUser } from 'src/types/user';
import { getLoggedUserId } from './users';

export const sortConversationsByLastActive = (conversations: Conversation[]): Conversation[] =>
  conversations.sort((a, b) => b.lastMessageTimestamp - a.lastMessageTimestamp);

export const getOtherUserFromConversation = (conversation: Conversation): BareUser =>
  conversation.senderId === getLoggedUserId()
    ? { id: conversation.recipientId, nickname: conversation.recipientNickname }
    : { id: conversation.senderId, nickname: conversation.senderNickname };
