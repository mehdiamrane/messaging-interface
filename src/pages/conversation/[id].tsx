import React, { useState, useEffect, useCallback, useRef } from 'react';
import type { FC } from 'react';
import Head from 'next/head';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import { GetServerSideProps } from 'next';

import dynamic from 'next/dynamic';

const ErrorBanner = dynamic(() => import('src/components/layout/errorBanner/ErrorBanner'), { ssr: false });

import { Conversation } from 'src/types/conversation';
import { Message } from 'src/types/message';

import { useTranslation } from 'next-i18next';
import Aside from 'src/components/layout/aside/Aside';
import Main from 'src/components/layout/main/Main';
// import ErrorBanner from 'src/components/layout/errorBanner/ErrorBanner';
import List from 'src/components/conversation/list/List';
import MessagesBox from 'src/components/conversation/messagesBox/MessagesBox';

import { getMessagesByConversationId, getConversationsByUserId } from 'src/services/api';
import { getLoggedUserId } from 'src/utils/users';
import { getOtherUserFromConversation } from 'src/utils/conversations';
import { BareUser } from 'src/types/user';

interface ConversationProps {
  conversations: Conversation[];
  currentConversation: Conversation;
  otherUser: BareUser;
}

const ConversationPage: FC<ConversationProps> = ({ conversations, currentConversation, otherUser }) => {
  const { t } = useTranslation('common');
  const [messages, setMessages] = useState<Message[]>();
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string>();

  const messagesEndRef = useRef(null);

  useEffect(() => {
    return () => {
      console.log('unmount');
      setLoading(true);
    };
  }, [currentConversation]);

  const getMessages = useCallback(async () => {
    console.log('getting messages');
    const { data: messagesData } = await getMessagesByConversationId(currentConversation.id);
    setMessages(messagesData);
    setLoading(false);
  }, [currentConversation]);

  const scrollToBottom = useCallback(() => {
    messagesEndRef.current.scrollIntoView({ behavior: 'smooth' });
  }, []);

  useEffect(() => {
    getMessages();
  }, [getMessages]);

  useEffect(() => {
    scrollToBottom();
  }, [messages, scrollToBottom]);

  const handleError = useCallback((error?: string) => {
    setError(error);
  }, []);

  return (
    <>
      <Head>
        <title>{t('conversation.meta.title', { name: otherUser.nickname })}</title>
        <meta name='description' content={t('conversation.meta.description', { name: otherUser.nickname })}></meta>
      </Head>
      <ErrorBanner error={error} onClose={() => setError(undefined)} />
      <Aside hideOnMobile>
        <List conversations={conversations} />
      </Aside>
      <Main hideOnMobile={false}>
        <MessagesBox
          fetchMessages={getMessages}
          handleError={handleError}
          loading={loading}
          messages={messages}
          conversation={currentConversation}
          messagesEndRef={messagesEndRef}
        />
      </Main>
    </>
  );
};

export default ConversationPage;

export const getServerSideProps: GetServerSideProps = async ({ params, locale }) => {
  const { id } = params;

  const { data: conversations } = await getConversationsByUserId(getLoggedUserId());

  const currentConversation = conversations.find((conv) => conv.id === Number(id));

  const otherUser = getOtherUserFromConversation(currentConversation);

  return {
    props: {
      otherUser,
      conversations,
      currentConversation,
      ...(await serverSideTranslations(locale, ['common'])),
    },
  };
};
