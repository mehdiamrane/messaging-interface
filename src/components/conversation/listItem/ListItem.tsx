import Image from 'next/image';
import React, { FC, useMemo } from 'react';
import { StyledListItem } from './ListItem.style';
import { Conversation } from 'src/types/conversation';
import { useTranslation } from 'next-i18next';
import { getAvatarFromNickname } from 'src/utils/users';
import { getOtherUserFromConversation } from 'src/utils/conversations';
import NextLink from 'next/link';
import { formatRelativeDate, timestampToDate } from 'src/utils/dates';
import { useRouter } from 'next/router';

interface ListItemProps {
  conversation: Conversation;
}

const ListItem: FC<ListItemProps> = ({ conversation }) => {
  const { t } = useTranslation('common');
  const router = useRouter();

  const { lastMessageTimestamp } = conversation;
  const formatedRelativeDate = useMemo(
    () => formatRelativeDate(timestampToDate(lastMessageTimestamp), router.locale),
    [lastMessageTimestamp, router],
  );
  const { nickname } = useMemo(() => getOtherUserFromConversation(conversation), [conversation]);
  const avatarSrc = useMemo(() => getAvatarFromNickname(nickname), [nickname]);

  const isCurrent = useMemo(() => conversation.id === Number(router?.query?.id), [conversation, router]);

  return (
    <NextLink passHref href={`/conversation/${conversation.id}`}>
      <StyledListItem.Container title={t('list.item.title', { name: nickname })} $isCurrent={isCurrent}>
        <Image src={avatarSrc} alt={t('list.item.avatar.alt', { name: nickname })} width={54} height={54} />

        <StyledListItem.TextWrapper>
          <StyledListItem.Name>{nickname}</StyledListItem.Name>
          <StyledListItem.Info>{formatedRelativeDate}</StyledListItem.Info>
        </StyledListItem.TextWrapper>
      </StyledListItem.Container>
    </NextLink>
  );
};

export default ListItem;
