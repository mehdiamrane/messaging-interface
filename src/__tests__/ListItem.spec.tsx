import React from 'react';
import { render } from 'jest/utils';
import { cleanup } from '@testing-library/react';
import ListItem from 'src/components/conversation/listItem/ListItem';
import 'jest-styled-components';
import { Conversation } from 'src/types/conversation';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
afterAll(cleanup);

describe('ListItem', () => {
  it('should render correctly ListItem', () => {
    useRouter.mockImplementation(() => ({
      query: {
        id: 1,
      },
    }));

    const dummyConversation: Conversation = {
      id: 1,
      recipientId: 2,
      recipientNickname: 'Jeremie',
      senderId: 1,
      senderNickname: 'Thibaut',
      lastMessageTimestamp: 1625637849,
    };

    const { container } = render(<ListItem conversation={dummyConversation} />);
    expect(container).toMatchSnapshot();
  });
});
