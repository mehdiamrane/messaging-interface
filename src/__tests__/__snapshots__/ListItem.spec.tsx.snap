// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should render correctly ListItem 1`] = `
.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  color: #4A5568;
  font-size: 1rem;
  font-weight: 500;
  -webkit-transition: color 150ms ease;
  transition: color 150ms ease;
}

.c5 {
  color: #718096;
  font-size: 0.875rem;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #EDF2F7;
  border-bottom: 1px solid #CBD5E0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5rem;
  gap: 1rem;
  padding: 0 1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 150ms ease;
  transition: background-color 150ms ease;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #90cdf4;
}

.c0 .c2,
.c0 .c4 {
  color: #2b6cb0;
}

<div>
  <a
    class="c0"
    href="/conversation/1"
    title="list.item.title"
  >
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2754%27%20height=%2754%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        alt="list.item.avatar.alt"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
    <div
      class="c1"
    >
      <span
        class="c2 c3"
      >
        Jeremie
      </span>
      <span
        class="c2 c4 c3 c5"
      >
        10 months ago
      </span>
    </div>
  </a>
</div>
`;
